### PR TITLE
Python 3 compatibility fixes

### PR DESCRIPTION
**Pull request by jdeschenes**
The unit test pass in python 3.6. **This has not yet been completely tested in python**. I will do another pass once I have gone over enable.

The examples have not been run as well. So I may make further modifications.

There is an issue in python 3 in the image plot. This is caused by a bug in enable. I skipped those tests.

### CODE REVIEW

File: chaco/legend.py [hunk 1]
```diff
@@ -355,7 +358,7 @@ def get_preferred_size(self):
         if len(self.plots) == 0:
             return [0, 0]
 
-        plot_names, visible_plots = map(list, zip(*sorted(self.plots.items())))
+        plot_names, visible_plots = list(sm.map(list, sm.zip(*sorted(six.iteritems(self.plots)))))
```
Comment (jvkersch) [inline chaco/legend.py]: This could probably do with some unpacking. Alternatively, can we just use the standard Python constructs here and write the RHS as `list(map(list, zip(*sorted(self.plots))))`?

File: chaco/plotscrollbar.py [hunk 2]
```diff
@@ -1,4 +1,6 @@
```
Comment (jvkersch) [inline chaco/plotscrollbar.py]: Can you put a `from __future__ import print_function` here?

File: chaco/scales/formatters.py [hunk 3]
```diff
@@ -350,15 +354,19 @@ def estimate_width(self, start, end, numlabels=None, char_width=None,
             elif char_width:
                 avg_size = len("%g%g" % (start, end)) / 2.0
                 initial_estimate = round(fill_ratio * char_width / avg_size)
+            else:
```
Comment (jvkersch) [inline chaco/scales/formatters.py]: This seems orthogonal to the cleanup, can you describe what the issue is here? If the issue is that `numlabels` and `char_width` should not be both `None`, perhaps a `ValueError("num_labels and char_width should not both be None.")` would be more appropriate.

File: chaco/scales/scales_test_case.py [hunk 4]
```diff
@@ -3,8 +3,8 @@
```
Comment (jvkersch) [inline chaco/scales/scales_test_case.py]: Also needs a `from __future__ import print_function`.

File: chaco/scales/tests/test_formatters.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def test_strftimeEx_04():
     # The format "%(ms_)" uses floor().
```
Comment (jvkersch) [inline chaco/scales/tests/test_formatters.py]: Also needs a `from __future__ import print_function` at the top of the module.

File: chaco/scales/time_scale_test_case.py [hunk 6]
```diff
@@ -2,11 +2,11 @@
 from itertools import starmap
```
Comment (jvkersch) [inline chaco/scales/time_scale_test_case.py]: Also needs a `from __future__ import print_function` at the top of the module.

File: chaco/tests/test_image_plot.py [hunk 7]
```diff
@@ -108,48 +113,56 @@ def plot_comparison(input_image, expected_image, **plot_kwargs):
     plt.show()
 
 
+@unittest.skipIf(six.PY3, "Bug in the image plotter in python 3")
```
Comment (jvkersch) [inline chaco/tests/test_image_plot.py]: That's unexpected, is that new?

File: chaco/tools/line_inspector.py [hunk 8]
```diff
@@ -78,7 +78,8 @@ def draw(self, gc, view_bounds=None):
         if self.is_listener:
             tmp = self._get_screen_pts()
         elif self.is_interactive:
-            tmp = self._last_position
```
Comment (jvkersch) [inline chaco/tools/line_inspector.py]: It looks like this (and some of the other changes in this module) changes the functionality. Can you explain why this is necessary? If this fixes a bug, it would probably be good to split it off from the current PR and submit as a separate PR.

File: chaco/tools/pan_tool.py [hunk 9]
```diff
@@ -247,6 +247,8 @@ def panning_mouse_leave(self, event):
         return self._end_pan(event)
 
     def _start_pan(self, event, capture_mouse=True):
+        if event.control_down:
```
Comment (jvkersch) [inline chaco/tools/pan_tool.py]: Also a functionality change.

File: docs/source/sphinxext/comment_eater.py [hunk 10]
```diff
@@ -1,10 +1,11 @@
-from cStringIO import StringIO
```
Comment (jvkersch) [inline docs/source/sphinxext/comment_eater.py]: I think some of the bundled sphinx extensions in this folder might not be Python3 compatible... (Just a comment in passing, no need to take any action)

File: examples/demo/basic/image_lasso.py [hunk 11]
```diff
@@ -27,15 +27,15 @@ def lasso_updated(lasso_tool, name, old, new_selections):
...
+        #print("Selection region", i)
 
         # We first map to screen because the selection is stored as coordinates
         # in data space
         screen_pts = lasso_tool.plot.map_screen(selection)
 
         # Now map each point into the grid index
-        for x, y in screen_pts:
-            print("\t", lasso_tool.plot.map_index((x, y)))
+        # for x, y in screen_pts:
```
Comment (jvkersch) [inline examples/demo/basic/image_lasso.py]: Can you remove the commented-out code here?

File: examples/demo/image_plot_origin_and_orientation.py [hunk 12]
```diff
@@ -13,7 +13,7 @@
 """
 
 # Major library imports
-from scipy.misc import lena
+from scipy.misc import face
```
Comment (jvkersch) [inline examples/demo/image_plot_origin_and_orientation.py]: +100

File: examples/demo/multiaxis_using_Plot.py [hunk 13]
```diff
@@ -24,6 +24,60 @@
 from chaco.tools.api import BroadcasterTool, PanTool, ZoomTool
 from chaco.api import create_line_plot, add_default_axes
 
+from traits.api import List
+from chaco.tools.api import SelectedZoomState
+import numpy
+
+class CaeZoomTool(ZoomTool):
```
Comment (jvkersch) [inline examples/demo/multiaxis_using_Plot.py]: Cae?

**Comment by jdeschenes** [comment]
Sorry about that, looks like I bundled more changes(I am using this branch) than strictly necessary. I will do a pass and remove functional changes.

File: chaco/tools/line_inspector.py (hunk 8 above)
Comment (jdeschenes) [inline chaco/tools/line_inspector.py]: Hi, I pulled in my branch #205 as I found it helped quite a bit and it looked like it will not be implemented anytime soon. I reverted changes in latest commit.

File: chaco/tools/pan_tool.py (hunk 9 above)
Comment (jdeschenes) [inline chaco/tools/pan_tool.py]: Sorry about that... This has been reverted.

File: examples/demo/multiaxis_using_Plot.py (hunk 13 above)
Comment (jdeschenes) [inline examples/demo/multiaxis_using_Plot.py]: Sorry about that, this should not have been part of the commit.

Comment (jvkersch) [inline examples/demo/multiaxis_using_Plot.py]: No worries, looked like a pretty nifty tool ;-)

**Comment by jdeschenes** [comment]
I think I need to comment on some of the functional changes that I put there by mistake.

For multiaxis_using_Plot.py:

1. Using the zoom tool with ctrl is doing a pan at the same time(pan_tool.py.) Strangely, this doesn't happen in two_plots.py. I think a code change may be necessary here. My approach was to subclass PanTool and ignore the pan move if my modifiers were pressed.
2. The Box on the ZoomTool does not appear
3. The ZoomTool is not doing the correct zoom(See issue #366) The CaeZoomTool has this fixed, but only partially... Using the mouse wheel to zoom off will not do what is right.

For financial_plot.py

1. The font does not render properly on my machine. There is a warning on font_manager.py in kiva. That's a complete separate issue.

These changes definitively don't belong in this pull request.

File: chaco/tests/test_image_plot.py (hunk 7 above)
Comment (jdeschenes) [inline chaco/tests/test_image_plot.py]: That's actually a pretty old [issue](https://github.com/enthought/enable/issues/283)

File: chaco/legend.py (hunk 1 above)
Comment (jdeschenes) [inline chaco/legend.py]: Your approach will not work. You need to have the items() in there

```python
x = {'b': 1, 'a': 3}

assert sorted(x) == ['a', 'b']
assert sorted(x.items()) == [('a', 3), ('b', 1)]
```

For standard python contructs. See my comment in enthought/traits#374

It's more or less the same issue. I would prefer to keep the iterator version whenever possible.

Let me know your decision.

Comment (jvkersch) [inline chaco/legend.py]: Ah, I misunderstood. Thanks for the clarification, and let's keep it like you had it originally.

File: chaco/tests/test_image_plot.py (hunk 7 above)
Comment (jvkersch) [inline chaco/tests/test_image_plot.py]: I see, can you add a reference to the issue in the skip message?

**Comment by jvkersch** [comment]
@jdeschenes Thanks for addressing the comments. Would you mind merging master in and resolving any merge conflicts? It would be good to start getting Travis test results on this.

**Comment by jdeschenes** [comment]
The merge has just been done.

File: chaco/speedups.py [hunk 14]
```diff
@@ -1,12 +1,14 @@
...
+from ._speedups_fallback import *
 
 
 # cython implementation of speedups. Import these if we can.
 try:
-    from _cython_speedups import *
+    from ._cython_speedups import *
 except ImportError:
+    # TO TEST
+    raise
```
Comment (jvkersch) [inline chaco/speedups.py]: Do we need this explicit `raise`?

File: docs/source/conf.py [hunk 15]
```diff
@@ -52,7 +52,9 @@ def get_build_docset():
 # The default replacements for |version| and |release|, also used in various
 # other places throughout the built documents.
 d = {}
-execfile(os.path.join('..', '..', 'chaco', '_version.py'), d)
+exec(compile(open(os.path.join('..', '..', 'chaco', '_version.py')).read(),
```
Comment (jvkersch) [inline docs/source/conf.py]: Can you isolate the `os.path.join('..', '..', 'chaco', '_version.py')` into a local variable for re-use?

**Comment by jvkersch** [comment]
@jdeschenes I left two tiny comments, but otherwise this looks very good to me. If you fix up those two items (and have nothing more to add), then I think this is good to go.

**Comment by jdeschenes** [comment]
@jvkersch I fixed the issues you mentionned. I think this is ready to go!

**Comment by jvkersch** [comment]
@jdeschenes I'm following along with the Enable review, and I think it would be good here too to use `.items()`, `.values()`, etc. instead of `six.iteritems(...)`, `six.itervalues(...)`, and so on. Would you be able to make that change? If not, I will pick this up separately (and apologies for bringing this up at the last minute).

**Comment by jdeschenes** [comment]
@jvkersch I think I have corrected everything. Let me know if there is anything else!

**Comment by jvkersch** [comment]
@jdeschenes I think this looks good to go! 

Thanks a lot for doing this (and for tackling the other repos, too). This will improve developer experience by a lot!
